lib/Stomp/Server.pm6

```diff
@@ -1,26 +1,110 @@
                     {
-                        await $conn.print: Stomp::Message.new:
+                        await $!conn.print: Stomp::Message.new:
                             command => 'CONNECTED',
                             headers => ( accept-version => '1.2' );
-                    },
-                    quit => {
-                        when X::Stomp::MalformedMessage {
-                            await $conn.print: Stomp::Message.new:
-                                command => 'ERROR',
-                                body => .message;
-                        }
-                    };
+                        $connect-tap.close;
+                    }, :&quit;
+            $!messages.grep({ $_.command ~~ 'SUBSCRIBE' }).tap: {
+                $!subscription-lock.protect: {
+                    @!subscriptions.push: Subscription.new( id          => $_.headers<id>,
+                                                            destination => $_.headers<destination>,
+                                                            ack         => $_.headers<ack> ?? AckMode($_.headers<ack>) !! Auto );
+                };
+            }, :&quit;
+            $!messages.grep({ $_.command ~~ 'UNSUBSCRIBE' }).tap: {
+                $!subscription-lock.protect: {
+                    my $id = $_.headers<id>;
+                    @!subscriptions = @!subscriptions.grep({ $_.id !~~ $id });
+                };
+            }, :&quit;
+
+            $!sent-message-supplier = Supplier.new;
+            $!messages.grep({ $_.command ~~ 'SEND' }).tap: {
+                $!sent-message-supplier.emit: $_;
+            }, :&quit;
+        }
+
+        method published-messages() returns Supply {
+            $!sent-message-supplier.Supply;
+        }
+
+        method ACCEPTS(Stomp::Message $mess) {
+            $!subscription-lock.protect({
+                @!subscriptions.map(*.destination).any eq $mess.headers<destination>;
+            });
+        }
+
+        method subscription-for-message(Stomp::Message $mess) {
+            $!subscription-lock.protect({
+                @!subscriptions.first({ $_.destination eq $mess.headers<destination>});
+            });
+        }
+
+        method send(Stomp::Message $mess) {
+            if self.subscription-for-message($mess) -> $sub {
+                my $subscription = $sub.id;
+                my $message-id = $mess.uuid;
+                my $out = Stomp::Message.new(
+                    command => 'MESSAGE',
+                    body    => $mess.body,
+                    headers => (|$mess.headers, :$message-id, :$subscription, )
+                );
+                $!conn.print: $out;
+            }
+            else {
+                fail "Not subscribed to message";
+            }
+        }
+    }
+
+    method listen() {
+        supply {
+            whenever self.socket-provider.listen($!host, $!port) -> $conn {
+                my $connection = Connection.new(:$conn);
```

Comment:
I'm wondering whether the Connection should have a Promise that is kept when the CONNECT has been completed (which lets us do the authentication or whatever as well,) so that it's not  emitted before we have a valid connection.

Reply:
I've implemented that simply.  I think there needs to be a hook for authentification.